File: frontend/lib/api/intelligence.ts
export type IntelligenceMeta = {
  methodology_version: string;
  as_of_block: number;
  computed_at: string;
  limitations: string[];
};

export type ValidatorRanking = {
  hotkey: string;
  subnet_id: number;
  composite_score: number;
  apy_estimate: number;
  reward_consistency: number;
  uptime_percent: number;
  rank_subnet: number;
  rank_global: number;
  delegation_trend: number;
  reputation_signal: number;
  meta: IntelligenceMeta;
};

export type ValidatorIntelligence = ValidatorRanking & {
  inputs: Record<string, number>;
};

export type PaginatedRankings = {
  data: ValidatorRanking[];
  pagination: {
    page: number;
    page_size: number;
    total_items: number;
    total_pages: number;
  };
};

export type WalletRiskProfile = {
  wallet_address: string;
  concentration_validator: number;
  concentration_subnet: number;
  hhi_validator: number;
  hhi_subnet: number;
  reward_volatility: number;
  downtime_risk_proxy: number;
  overall_risk_band: string;
  inputs: Record<string, number>;
  meta: IntelligenceMeta;
};

export type RewardForecast = {
  wallet_address: string;
  methodology_version: string;
  limitations: string[];
  is_estimate: boolean;
  implied_apy_pct: number;
  history_days: number;
  forecast: Array<{ day_offset: number; amount_rao: number }>;
};

export type CompareValidators = {
  validators: ValidatorIntelligence[];
  limitations: string[];
};

File: backend/app/api/v1/schemas/intelligence.py
from datetime import datetime

from pydantic import BaseModel, Field


class IntelligenceMeta(BaseModel):
    methodology_version: str
    as_of_block: int
    computed_at: datetime
    limitations: list[str]


class ValidatorRankingResponse(BaseModel):
    hotkey: str
    subnet_id: int
    composite_score: int
    apy_estimate: float
    reward_consistency: float
    uptime_percent: float
    rank_subnet: int
    rank_global: int
    delegation_trend: float
    reputation_signal: float
    meta: IntelligenceMeta


class ValidatorIntelligenceResponse(ValidatorRankingResponse):
    inputs: dict[str, float]


class WalletRiskResponse(BaseModel):
    wallet_address: str
    concentration_validator: float
    concentration_subnet: float
    hhi_validator: float
    hhi_subnet: float
    reward_volatility: float
    downtime_risk_proxy: float
    overall_risk_band: str
    inputs: dict[str, float]
    meta: IntelligenceMeta


class ForecastPoint(BaseModel):
    day_offset: int
    amount_rao: int


class RewardForecastResponse(BaseModel):
    wallet_address: str
    methodology_version: str
    limitations: list[str]
    is_estimate: bool = True
    implied_apy_pct: float
    history_days: int
    forecast: list[ForecastPoint]


class CompareValidatorsResponse(BaseModel):
    validators: list[ValidatorIntelligenceResponse]
    limitations: list[str]
